waspc/data/Generator/templates/react-app/src/actions/index.js

```diff
@@ -0,0 +1,74 @@
+import {
+  useMutation,
+  useQueryClient,
+} from 'react-query'
+
+export { configureQueryClient } from '../queryClient'
+
+export function useAction(actionFn, actionOptions) {
+  const queryClient = useQueryClient();
+
+  // TODO(filip): Do we allow all react query options or should we consider a list of allowed options?
+  let options = {}
+
+  if (actionOptions?.optimisticUpdates) {
+    const optimisticUpdatesConfig = makeOptimisticUpdatesConfig(actionOptions.optimisticUpdates)
+    options = makeOptimisticUpdateOptions(queryClient, optimisticUpdatesConfig)
+  }
+
+  return useMutation(actionFn, options)
+}
+
+
+function makeOptimisticUpdatesConfig(optimisticUpdatesConfig) {
+  return optimisticUpdatesConfig.map(({ getQuery, ...rest }) => ({
+    getQuery: (item) => parseQueryKey(getQuery(item)),
+    ...rest,
+  }))
+}
+
+function makeOptimisticUpdateOptions(queryClient, optimisticUpdatesConfig) {
```

Comment:
Could you pls comment this whole function, both regarding external API (what it takes / does / returns) and regarding its internal steps?
I have to admit that I am finding JS very hard to read and understand since there are no types + the stuff we are doing here is relatively complex so I am missing a lot of context -> if you could provide some of it via comments, that would help me a lot in understanding this code.

Reply:
Definitely, will do.